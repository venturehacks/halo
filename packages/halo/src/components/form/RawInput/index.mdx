---
name: RawInput
menu: Form
route: /form/raw-input
---

import { Props, Playground } from 'docz';
import { Label } from '../Label';
import { RawInput } from './';
import { FieldErrorMessage } from '../FieldErrorMessage';
import { TimeIcon, ArchiveIcon } from '../../icons';
import { Bracket } from '../../structure/Bracket';

# RawInput

Raw input fields are designed to be controlled by a [smart, form-connected parent](https://github.com/venturehacks/AngelList/blob/master/frontend/talent/components/Form/TextField/index.tsx).



### Usage

<Playground>
  <div className="max-w-xl">
    <Label field="foo" title="Text Input">
      <RawInput
        type="text"
        name="foo"
        placeholder="Your name"
        maxLength={100}
      />
    </Label>
    <Bracket>
      <Label field="foo" title="Icon Left">
        <RawInput
          type="text"
          name="foo"
          placeholder="Timezone name..."
          maxLength={100}
          icon={<TimeIcon />}
        />
      </Label>
      <Label field="foo" title="Icon Right">
        <RawInput
          type="text"
          name="foo"
          placeholder="archive@foo.com"
          maxLength={100}
          iconPosition="right"
          icon={<ArchiveIcon />}
        />
      </Label>
    </Bracket>
    <Bracket>
      <Label field="bar" title="Email">
        <RawInput
          type="email"
          name="bar"
          onChange={() => {}}
          value="What is email?"
          intrinsicWidth="auto"
          hasError
        />
        <FieldErrorMessage
          errorSeverity="warning"
          message="An email is required"
        />
      </Label>
      <Label field="password" title="Password">
        <RawInput
          type="password"
          name="password"
          intrinsicWidth="auto"
          value="abcdefghijk"
          onChange={() => {}}
          errorSeverity="critical"
          hasError
        />
        <FieldErrorMessage
          errorSeverity="critical"
          message="Does not match"
        />
      </Label>
    </Bracket>
    <Label field="phone" title="Number">
      <RawInput
        type="number"
        name="phone"
        intrinsicWidth="auto"
        min="10"
        max="100"
        value="20"
        onChange={() => {}}
      />
    </Label>
  </div>
</Playground>

### Props

<Props of={RawInput} />
