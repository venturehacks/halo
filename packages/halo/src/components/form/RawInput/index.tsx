import classNames from 'classnames';
import React from 'react';

import {
  FORM_FIELD_ERROR_IDENTIFIER,
  ForwardedRefProps,
  withForwardedRef,
} from '../../../lib';
import { Box } from '../../structure/Box';

import styles from './styles.scss';

type IconPosition = 'right' | 'left';

export interface RawInputProps {
  className?: string;

  /**
   * Since the most common callout is for validation errors, you shouldn't need to customize this
   * @default warning
   */
  errorSeverity?: 'warning' | 'critical';
  /**
   * Call out element that needs attention
   * @default false
   */
  hasError?: boolean;

  /**
   * Optional prefix/suffix icon
   */
  icon?: React.ReactNode;

  /**
   * className applied to icon container div
   */
  iconContainerClassName?: string;

  /**
   * left | right
   * @default left
   */
  iconPosition?: IconPosition;

  /**
   * Grows as wide as container by default
   * @default 100%
   */
  intrinsicWidth?: 'auto' | '100%';

  /**
   * Field name
   */
  name?: string;

  onChange?: React.ChangeEventHandler<HTMLInputElement> | (() => void);

  /**
   * Use transparent style
   * @default false
   */
  transparent?: boolean;
  /**
   * Standard HTML <input> element types
   * @default text
   */
  type?:
    | 'date'
    | 'datetime-local'
    | 'email'
    | 'month'
    | 'number'
    | 'password'
    | 'search'
    | 'tel'
    | 'text'
    | 'time'
    | 'url'
    | 'week';
}

function RawInputRaw({
  className,
  errorSeverity = 'warning',
  forwardedRef,
  hasError = false,
  icon,
  iconContainerClassName,
  iconPosition = 'left',
  intrinsicWidth = '100%',
  transparent = false,
  type = 'text',
  ...rest
}: RawInputProps &
  ForwardedRefProps<HTMLInputElement> &
  React.InputHTMLAttributes<HTMLInputElement>) {
  const input = (
    <input
      ref={forwardedRef}
      className={classNames(
        styles.component,
        className,
        transparent ? styles.transparent : styles.bordered,
        icon && iconPosition === 'right' && styles.hasIconRight,
        icon && iconPosition === 'left' && styles.hasIconLeft,
        intrinsicWidth === 'auto' && styles.widthAuto,
        hasError && styles.hasError,
        hasError && FORM_FIELD_ERROR_IDENTIFIER,
        errorSeverity === 'warning' && styles.warning,
        errorSeverity === 'critical' && styles.critical,
      )}
      type={type}
      {...rest}
    />
  );

  if (!icon) {
    return input;
  }

  return (
    <div className={styles.wrapper}>
      {input}
      <Box
        align="center"
        className={classNames(
          styles.iconContainer,
          iconPosition === 'right' && styles.hasIconRight,
          iconContainerClassName,
        )}
        valign="center"
        column
      >
        {icon}
      </Box>
    </div>
  );
}

// NOTE(drew): I wonder if we need to go back to `extends ...`
const RawInput = withForwardedRef<
  RawInputProps & React.InputHTMLAttributes<HTMLInputElement>,
  HTMLInputElement
>(RawInputRaw);

export { RawInput };
