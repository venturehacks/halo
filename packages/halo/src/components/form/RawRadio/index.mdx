---
name: RawRadio
menu: Form
route: /form/raw-radio
---

import { Props, Playground } from 'docz';
import { Label } from '../Label';
import { RawRadio } from './';

# RawRadio

Raw input fields are designed to be controlled by a [smart, form-connected parent](https://github.com/venturehacks/AngelList/blob/master/frontend/talent/components/Form/TextField/index.tsx).

### Usage

<Playground>
  <Label field="ice-cream" title="One True Ice Cream" byline="What's your favorite?" containsFieldGroup>
    <div className="flex flex-row gap-10 max-w-xl mb-8">
      <RawRadio
        name="ice-cream"
        value="chocolate"
        label="Chocolate"
        onChange={() => {}}
      />
      <RawRadio
        name="ice-cream"
        value="vanilla"
        label="Vanilla"
        onChange={() => {}}
        checked
      />
    </div>
  </Label>
  <Label field="disabled-ice-cream" title="Disabled" byline="Sold out" containsFieldGroup>
    <div className="flex flex-row gap-10 max-w-xl mb-8">
      <RawRadio
        name="disabled-ice-cream"
        value="banana"
        label="Banana"
        onChange={() => {}}
        disabled
      />
      <RawRadio
        name="disabled-ice-cream"
        value="rocky road"
        label="Rocky Road"
        onChange={() => {}}
        checked
        disabled
      />
    </div>
  </Label>
  <Label field="another-ice-cream" title="With errors" byline="not quite sprinkles" containsFieldGroup>
    <div className="flex flex-row gap-10 max-w-xl">
      <RawRadio
        name="another-ice-cream"
        value="caramel"
        label="Caramel"
        onChange={() => {}}
        hasError
      />
      <RawRadio
        name="another-ice-cream"
        value="brownie"
        label="Brownie"
        onChange={() => {}}
        errorSeverity="critical"
        checked
        hasError
      />
    </div>
  </Label>
</Playground>

### Props

<Props of={RawRadio} />
