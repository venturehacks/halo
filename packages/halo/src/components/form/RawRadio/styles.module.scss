@import 'halo';

.input {
  // hide-default-browser-input
  opacity: 0;
  position: absolute;
  -webkit-appearance: none;
  z-index: -2;

  &[disabled] {
    pointer-events: none;
  }
}

// Label - larger clickable area is also responsible for rendering checked state

.checkbox {
  @include input-label-checkbox;
}

.radio {
  @include input-label-radio;
}

.pill {
  @include input-label-pill;
}

.controlAlignCenter {
  @include input-control-center;
}
