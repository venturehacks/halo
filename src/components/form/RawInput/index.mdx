---
name: RawInput
menu: Form
route: /form/raw-input
---

import { Props, Playground } from 'docz';
import { Label } from '../Label';
import { RawInput } from './';
import { Box } from '../../structure/Box';

# RawInput

ðŸ“£ Raw input fields are designed to be [controlled](https://reactjs.org/docs/forms.html#controlled-components) by their smarter, form-connected parent.

This component is primarily used for simple text input. It's also great as a hidden field that holds the value of a complex frontend component. On CoPro, we use it to power our autocomplete text fields and tabular data filter fields.

### Usage

<Playground>
  <Box align="stretch" maxWidth="500px">
    <Label field="foo" title="Default Behavior">
      <RawInput />
    </Label>
    <Label field="foo" title="Text Input">
      <RawInput
        type="text"
        name="foo"
        placeholder="Your name"
        maxLength={100}
      />
    </Label>
    <Label field="bar" title="Email" supportingText="with validation error, intrinsic width = auto">
      <RawInput
        type="email"
        name="bar"
        onChange={() => {}}
        value="What is email?"
        intrinsicWidth="auto"
        hasError
      />
    </Label>
    <Label field="password" title="Password">
      <RawInput
        type="password"
        name="password"
        intrinsicWidth="auto"
        value="abcdefghijk"
        onChange={() => {}}
      />
    </Label>
    <Label field="phone" title="Number">
      <RawInput
        type="number"
        name="phone"
        intrinsicWidth="auto"
        min="10"
        max="100"
        value="20"
        onChange={() => {}}
      />
    </Label>
  </Box>
</Playground>

### Props

<Props of={RawInput} />
