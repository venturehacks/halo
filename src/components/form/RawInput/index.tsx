import classNames from 'classnames';
import * as React from 'react';

import { FORM_FIELD_ERROR_IDENTIFIER } from '../../../lib';
import { Box } from '../../structure/Box';

import * as styles from './styles.scss';

type IconPosition = 'right' | 'left';

export interface RawInputProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  className?: string;
  /**
   * Call out element that needs attention
   * @default false
   */
  hasError?: boolean;
  /**
   * Since the most common callout is for validation errors, you shouldn't need to customize this
   * @default warning
   */
  errorSeverity?: 'warning' | 'critical';
  /**
   * Optional prefix/suffix icon
   */
  icon?: React.ReactNode;
  /**
   * left | right
   * @default left
   */
  iconPosition?: IconPosition;
  /**
   * Grows as wide as container by default
   * @default 100%
   */
  intrinsicWidth?: 'auto' | '100%';
  /**
   * Use transparent style
   * @default false
   */
  transparent?: boolean;

  /**
   * Standard HTML <input> element types
   * @default text
   */
  type?:
    | 'date'
    | 'datetime-local'
    | 'email'
    | 'month'
    | 'number'
    | 'password'
    | 'search'
    | 'tel'
    | 'text'
    | 'time'
    | 'url'
    | 'week';
}

export default function RawInput({
  className,
  errorSeverity = 'warning',
  hasError = false,
  icon,
  iconPosition = 'left',
  intrinsicWidth = '100%',
  transparent = false,
  type = 'text',
  ...rest
}: RawInputProps) {
  const input = (
    <input
      className={classNames(
        styles.component,
        className,
        transparent ? styles.transparent : styles.bordered,
        icon && iconPosition === 'right' && styles.hasIconRight,
        icon && iconPosition === 'left' && styles.hasIconLeft,
        intrinsicWidth === 'auto' && styles.widthAuto,
        hasError && styles.hasError,
        hasError && FORM_FIELD_ERROR_IDENTIFIER,
        errorSeverity === 'warning' && styles.warning,
        errorSeverity === 'critical' && styles.critical,
      )}
      type={type}
      {...rest}
    />
  );

  if (!icon) {
    return input;
  }

  return (
    <div className={styles.wrapper}>
      {input}
      <Box
        valign="center"
        align="center"
        className={classNames(
          styles.iconContainer,
          iconPosition === 'right' && styles.hasIconRight,
        )}
      >
        {icon}
      </Box>
    </div>
  );
}

export { RawInput };
