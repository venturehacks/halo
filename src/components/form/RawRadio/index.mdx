---
name: RawRadio
menu: Form
route: /form/raw-radio
---

import { Props, Playground } from 'docz';
import { Label } from '../Label';
import { RawRadio } from './';
import { Box } from '../../structure/Box';

# RawRadio

Raw input fields are designed to be completely controlled by their smarter, form-connected parent. Be sure to provide an `onChange` handler to manually manage the state of `checked`.

<Playground>
  <Label field="ice-cream" name="One True Ice Cream" supportingText="What's your favorite?" containsFieldGroup>
    <Box row align="space-between" maxWidth="500px">
    <RawRadio
      name="ice-cream"
      value="chocolate"
      label="Chocolate"
      onChange={() => {;}}
    />
    <RawRadio
      name="ice-cream"
      value="vanilla"
      label="Vanilla"
      onChange={() => {;}}
      checked
    />
    <RawRadio
      name="ice-cream"
      value="strawberry"
      label="Strawberry"
      onChange={() => {;}}
    />
    </Box>
  </Label>
</Playground>

### Props

<Props of={RawRadio} />
