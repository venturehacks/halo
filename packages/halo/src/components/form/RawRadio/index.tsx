import classNames from 'classnames';
import React from 'react';

import { ControlAlignment, FORM_FIELD_ERROR_IDENTIFIER } from '../../../lib';
import { FormInputErrorSeverity } from '../RawInput';

import styles from './styles.module.scss';

export interface RawRadioProps
  extends React.InputHTMLAttributes<HTMLInputElement> {
  /**
   * set to true to visually render radio elements as checkboxes
   * @default false
   */
  asCheckbox?: boolean;
  className?: string;
  /**
   * Vertical alignment of radio element
   * @default top
   */
  controlAlignment?: ControlAlignment;
  /**
   * Validation error = warning
   * Server error = critical
   * @default warning
   */
  errorSeverity?: FormInputErrorSeverity;
  /**
   * Call out element that needs attention
   * @default false
   */
  hasError?: boolean;
  /**
   * Required for associated <label>
   */
  id: string;
  label: React.ReactNode;
  labelClassName?: string;
  onChange?: React.ChangeEventHandler<HTMLInputElement> | (() => void);
  variant?: 'plain' | 'pill';
}

function RawRadio({
  className,
  type,
  id,
  label,
  labelClassName,
  asCheckbox = false,
  variant = 'plain',
  controlAlignment = 'top',
  hasError,
  errorSeverity = 'warning',
  ...rest
}: RawRadioProps) {
  return (
    <>
      <input
        className={classNames(
          styles.input,
          className,
          hasError && styles.hasError,
          hasError && FORM_FIELD_ERROR_IDENTIFIER,
          errorSeverity === 'warning' && styles.warning,
          errorSeverity === 'critical' && styles.critical,
        )}
        id={id}
        type="radio"
        {...rest}
      />
      <label
        className={classNames(
          labelClassName,
          styles.label,
          asCheckbox ? styles.asCheckbox : styles.asRadio,
          variant === 'pill' && styles.pill,
          controlAlignment === 'center' && styles.controlAlignCenter,
        )}
        htmlFor={id}
      >
        {label}
      </label>
    </>
  );
}

RawRadio.displayName = 'RawRadio';

export { RawRadio };
