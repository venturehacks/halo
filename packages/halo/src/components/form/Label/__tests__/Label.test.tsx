import { render } from 'enzyme';
import React from 'react';

import { RawInput } from '../../RawInput';
import { RawRadio } from '../../RawRadio';
import { Label } from '../index';

describe('Label', () => {
  test('smoke', () => {
    const component = render(<Label byline="Who are you?" title="Name" />);

    expect(component.text()).toContain('Name');
    expect(component.text()).toContain('Who are you?');
  });

  test('field group without containsFieldGroup throws', () => {
    expect(() =>
      render(
        <Label byline="Who are you?" title="Name">
          <RawRadio
            id="chocolate"
            label="Chocolate"
            onChange={() => {}}
            checked
          />
          <RawRadio id="vanilla" label="Vanilla" />
          <RawRadio id="strawberry" label="Strawberry" />
        </Label>,
      ),
    ).toThrowError();
  });

  describe('snapshot', () => {
    test('standard', () => {
      const component = render(
        <Label byline="Who are you?" title="Name">
          <RawInput />
        </Label>,
      );
      expect(component).toMatchSnapshot();
    });

    test('containsFieldGroup', () => {
      const component = render(
        <Label byline="Who are you?" title="Name" containsFieldGroup>
          <RawRadio
            id="chocolate"
            label="Chocolate"
            onChange={() => {}}
            checked
          />
          <RawRadio id="vanilla" label="Vanilla" />
          <RawRadio id="strawberry" label="Strawberry" />
        </Label>,
      );
      expect(component).toMatchSnapshot();
    });
  });
});
