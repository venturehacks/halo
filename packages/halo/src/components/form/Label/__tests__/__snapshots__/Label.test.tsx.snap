// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label snapshot containsFieldGroup 1`] = `
<fieldset
  class="block mb-6 border-none"
  name="form-input--undefined"
>
  <legend
    class="mb-3 w-full"
  >
    <div
      class="mb-1"
    >
      <div
        class="text-dark-aaaa text-md font-medium antialiased"
      >
        Name
      </div>
      <div
        class="text-dark-a mb-2 text-sm"
      >
        Who are you?
      </div>
    </div>
  </legend>
  <input
    checked=""
    class="input warning"
    id="chocolate"
    type="radio"
  />
  <label
    class="radio"
    for="chocolate"
  >
    Chocolate
  </label>
  <input
    class="input warning"
    id="vanilla"
    type="radio"
  />
  <label
    class="radio"
    for="vanilla"
  >
    Vanilla
  </label>
  <input
    class="input warning"
    id="strawberry"
    type="radio"
  />
  <label
    class="radio"
    for="strawberry"
  >
    Strawberry
  </label>
</fieldset>
`;

exports[`Label snapshot standard 1`] = `
<label
  class="block mb-6"
  for="form-input--undefined"
>
  <div
    class="mb-1"
  >
    <div
      class="text-dark-aaaa text-md font-medium antialiased"
    >
      Name
    </div>
    <div
      class="text-dark-a mb-2 text-sm"
    >
      Who are you?
    </div>
  </div>
  <input
    class="text-dark-aaaa text-md max-w-full placeholder-dark-a ring-inset border rounded focus:ring-dark-link bg-white  focus:ring-1 px-3 py-2 h-10 border-gray-500 disabled:text-dark-a disabled:bg-gray-200 disabled:border-light-aa w-full"
    type="text"
  />
</label>
`;
